package/kedro_viz/data_access/repositories/modular_pipelines.py

```diff
@@ -1,246 +1,324 @@
+            inputs (Set[str]): The input datasets to add.
+        """
+        hashed_inputs = set()
 
-        is_internal_input = modular_pipeline_id in input_node.modular_pipelines
-        if is_internal_input:
-            self.tree[modular_pipeline_id].internal_inputs.add(input_node.id)
-        else:
-            self.tree[modular_pipeline_id].external_inputs.add(input_node.id)
+        for _input in inputs:
+            hashed_input = self._hash_input_output(_input)
+            hashed_inputs.add(hashed_input)
+            if is_dataset_param(_input):
+                self.parameters.add(hashed_input)
 
-    def add_output(self, modular_pipeline_id: str, output_node: GraphNode):
-        """Add an output to a modular pipeline based on whether it's an internal or external output.
-        The output has knowledge of which modular pipelines it belongs to based on its namespace.
-        The information can be accessed with through `output_node.modular_pipelines`.
+        self.tree[modular_pipeline_id].inputs = hashed_inputs
+
+    def add_outputs(self, modular_pipeline_id: str, outputs: Set[str]) -> None:
+        """
+        Add output datasets from the modular pipeline.
 
         Args:
-            modular_pipeline_id: ID of the modular pipeline to add the output to.
-            output_node: The output node to add.
-        Raises:
-            ValueError: when attempt to add a non-data, non-parameter node as output
-                to the modular pipeline.
-        Example:
-            >>> modular_pipelines = ModularPipelinesRepository()
-            >>> data_science_pipeline = modular_pipelines.get_or_create_modular_pipeline(
-            ...     "data_science"
-            ... )
-            >>> model_output_node = GraphNode.create_data_node(
-            ...     "data_science.model_output", layer=None, tags=set(), dataset=None
-            ... )
-            >>> modular_pipelines.add_output("data_science", model_output_node)
-            >>> assert data_science_pipeline.outputs == {model_output_node.id}
-        """
-        if not isinstance(output_node, (DataNode, TranscodedDataNode, ParametersNode)):
-            raise ValueError(
-                f"Attempt to add a non-data node as input to modular pipeline {modular_pipeline_id}"
-            )
+            modular_pipeline_id (str): The ID of the modular pipeline to add outputs to.
+            outputs (Set[str]): The output datasets to add.
+        """
+        hashed_outputs = {self._hash_input_output(output) for output in outputs}
+        self.tree[modular_pipeline_id].outputs = hashed_outputs
 
-        is_internal_output = modular_pipeline_id in output_node.modular_pipelines
-        if is_internal_output:
-            self.tree[modular_pipeline_id].internal_outputs.add(output_node.id)
-        else:
-            self.tree[modular_pipeline_id].external_outputs.add(output_node.id)
+    def _hash_input_output(self, item: str) -> str:

```

Comment:
Why does this need to take `self`? As far as I understand this method just hashes input/output so it doesn't need an instance of `ModularPipelinesRepository`

Reply:
Yes, I would like to know where should the helper functions like this be placed in ? I used to place all the helper functions in utils as public. But there should be some rules which I am not aware of. Could you please let me know ? Thanks

Placed in -

1. A class as static method (public/private)
2. Outside of a class but in same file (public/private)
3. Inside utils (public/private)

Reply:
This stackoverflow thread, specifically the first two answers, describe quite well what the general rule of thumb is when it comes to static methods, class methods and module functions. https://stackoverflow.com/questions/11788195/module-function-vs-staticmethod-vs-classmethod-vs-no-decorators-which-idiom-is

In the Kedro codebase we generally apply the following:
1. Inside a class as static method: any method that is only used by that class, but the difference with a module function is that these types of methods are specific to this class and don't have any meaning outside of it. (e.g. https://github.com/kedro-org/kedro/blob/main/kedro/io/data_catalog.py#L336-L338)
2. Outside of a class but in same file: any helper method that is only used in that specific file should be a module function at the top of the file. Note that these methods are only used in that file/class, but don't require any "knowledge" about the class: no need to access a class instance and the method has meaning outside of the context of the class (e.g. https://github.com/kedro-org/kedro/blob/main/kedro/framework/session/session.py#L33-L56) 
3. Inside utils: when it's a helper method used in multiple files and not specific to a certain class. 

When it comes to making things public/private it's a bit different for Viz vs. Kedro but in the framework anything that's _public_ is part of the public API. So it can (and will) be used by outside users directly accessing Kedro as a library. So that means, unless a user should have access to the method, we make it _private_. My personal opinion is that for Viz we should make things _public_ when it's accessed in multiple classes/files.

Reply:
Thank you for the explanation Merel 💯